sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala

```diff
@@ -2574,6 +2574,27 @@ object SQLConf {
       .booleanConf
       .createWithDefault(false)
 
+  val COALESCE_BUCKET_IN_JOIN_ENABLED =
+    buildConf("spark.sql.bucketing.coalesceBucketInJoin.enabled")
+      .internal()
+      .doc("When true, if two bucketed tables with a different number of buckets are joined, " +
+        "the side with a bigger number of buckets will be coalesced to have the same number " +
+        "of buckets as the other side. This bucket coalescing can happen only when the bigger " +
+        "number of buckets is divisible by the smaller number of buckets.")
+      .version("3.1.0")
+      .booleanConf
+      .createWithDefault(false)
+
+  val COALESCE_BUCKET_IN_JOIN_MAX_NUM_BUCKETS_DIFF =
+    buildConf("spark.sql.bucketing.coalesceBucketInJoin.maxNumBucketsDiff")
```

Comment:
nit: How about `maxNumBucketsDiff ` -> `maxNumBucketsRatio`?

Reply:
The reason I picked a diff is we could have ratio of 2, but the actual difference could be large (e.g, 1000 / 500 => ratio of 2 but diff is 500).